src/crafting.cpp

```diff
@@ -993,82 +993,147 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
+    total_skill_modifiers += int_cur / 4.0f;
 
-    int diff_dice;
-    if( secondary_difficulty > 0 ) {
-        diff_dice = making.difficulty * 3 + secondary_difficulty;
-    } else {
-        // Since skill level is * 4 also
-        diff_dice = making.difficulty * 4;
+    // Missing proficiencies penalize skill level
+    // At the time of writing this is currently called a fail multiplier.
+    // TK: change the name of this feature to "skill penalty" or "skill bonus"
+    for( const recipe_proficiency &recip : making.proficiencies ) {
+        if( !recip.required && !has_proficiency( recip.id ) ) {
+            total_skill_modifiers -= ( recip.fail_multiplier - 1.0f );
+        }
     }
 
-    const int diff_sides = 24; // 16 + 8 (default intelligence)
+    add_msg_debug( debugmode::DF_CHARACTER, "Total skill modifiers: %f", total_skill_modifiers );
+    return weighted_skill_average + total_skill_modifiers;
+}
 
-    const double skill_roll = dice( skill_dice, skill_sides );
-    const double diff_roll = dice( diff_dice, diff_sides );
+double Character::crafting_success_roll( const recipe &making ) const
+{
+    // We're going to use a sqrt( sum of squares ) method here to give diminishing returns for more low level helpers.
+    float player_weighted_skill_average = get_recipe_weighted_skill_average( making );
+    bool negative = player_weighted_skill_average < 0;
+    player_weighted_skill_average = std::pow( player_weighted_skill_average, 2 );
+    if( negative ) {
+        player_weighted_skill_average *= -1;
+    }
+
+    float npc_helpers_weighted_average = 0.0f;
+    for( const npc *np : get_crafting_helpers() ) {
+        // can the NPC actually craft this recipe?
+        bool has_all_skills = np->get_skill_level( making.skill_used ) >= making.difficulty;
+        if( has_all_skills ) {
+            for( const std::pair<const skill_id, int> &secondary_skills : making.required_skills ) {
+                if( np->get_skill_level( secondary_skills.first ) < secondary_skills.second ) {
+                    has_all_skills = false;
+                }
+            }
+        }
+        if( has_all_skills && !making.character_has_required_proficiencies( *np ) ) {
+            has_all_skills = false;
+        }
+        if( has_all_skills ) {
+            const float helper_skill_average = std::max( np->get_recipe_weighted_skill_average( making ),
+                                               0.0f );
+            npc_helpers_weighted_average += std::pow( helper_skill_average, 2 );
+            add_msg_if_player( m_info, _( "%s helps with crafting…" ), np->name );
+        }
+    }
 
-    if( diff_roll == 0 ) {
-        // Automatic success
-        return 2;
+    // Use a sum of squares to determine the final weighted average.
+    negative = false;
+    float summed_skills = player_weighted_skill_average + npc_helpers_weighted_average;
+    if( summed_skills < 0 ) {
+        negative = true;
```

Comment:
```suggestion
        negative = true;
        // Store that it was negative to reapply later, then make it positive because we need a sqrt.
        summed_skills *= -1;
```
I was right the first time, math is hard.